### PR TITLE
Add projects model

### LINKED ISSUE #63: Add Project Model
Add model for Projects based on existing Persons model.

### PR DESCRIPTION
**Pull request by samuelfangjw**
Currently a draft model, waiting for tasks model to be completed.
Closes #63.

### CODE REVIEW

File: src/main/java/seedu/address/model/project/UniqueProjectList.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        internalList.set(index, editedProject);
+    }
+
+    /**
+     * Removes the equivalent project from the list.
+     * The project must exist in the list.
+     */
+    public void remove(Project toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/UniqueProjectList.java:80]: line space between if statement and previous line

File: src/main/java/seedu/address/model/project/UniqueProjectList.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the contents of this list with {@code projects}.
+     * {@code projects} must not contain duplicate projects.
+     */
+    public void setProjects(List<Project> projects) {
+        requireAllNonNull(projects);
+        if (!projectsAreUnique(projects)) {
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/UniqueProjectList.java:98]: line space

File: src/main/java/seedu/address/model/project/UniqueProjectList.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    @Override
+    public Iterator<Project> iterator() {
+        return internalList.iterator();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof UniqueProjectList // instanceof handles nulls
+                        && internalList.equals(((UniqueProjectList) other).internalList));
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/UniqueProjectList.java]: indentation. && should be aligned to ||.

**Review by vevek — APPROVED** [review]
Looks good to merge. Minor errors. I suggest adding comments to all the methods now so that it's clearer and won't be a bigger hassle in the future.

File: src/main/java/seedu/address/model/project/CompletableTaskList.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package seedu.address.model.project;
+
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CompletableTaskList {
+
+    private final List<Completable> completableTasks = new new ArrayList<>();
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/CompletableTaskList.java]: Do add the types in the array list as well. It's failing the checks.
"new" has been duplicated twice.

File: src/main/java/seedu/address/model/project/ParticipantList.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import seedu.address.model.person.Person;
+
+public class ParticipantList {
+
+    private final List<Person> participants = new ArrayList<>();
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/ParticipantList.java:12]: Here

**Review by vevek — CHANGES_REQUESTED** [review]
Need to make these minor changes too.

File: src/main/java/seedu/address/model/project/CompletableTaskList.java (hunk 4 above)
Comment (samuelfangjw) [reply src/main/java/seedu/address/model/project/CompletableTaskList.java]: Waiting for your PR @vevek, which would have implemented the required types.

**Comment by samuelfangjw** [comment]
> Looks good to merge. Minor errors. I suggest adding comments to all the methods now so that it's clearer and won't be a bigger hassle in the future.

Which methods are you referring to? There is no need for javadoc comments on getters and setters. Method with @Override tag inherit javadoc comments from superclass, and there is generally no need to add a javadoc comment unless the behaviour has been modified or there is something of note. e.g. equals, toString, hashcode.

File: src/main/java/seedu/address/model/project/CompletableTaskList.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * Constructs a empty {@code CompletableTaskList}.
+     */
+    public CompletableTaskList() {}
+
+    /**
+     * Constructs a {@code CompletableTaskList}.
+     *
+     * @param completableTasks A list of {@code Person}.
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/CompletableTaskList.java]: A list of Persons -> Tasks instead.

File: src/main/java/seedu/address/model/project/EventList.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public EventList () {}
+
+    /**
+     * Constructs an {@code EventList}.
+     *
+     * @param events A list of {@code Event}.
+     */
+    public EventList (List<Event> events) {
+        requireAllNonNull(events);
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/EventList.java]: This one can use requireNonNull(), but no big deal.

File: src/main/java/seedu/address/model/project/CompletableTaskList.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public CompletableTaskList(List<Completable> completableTasks) {
+        requireAllNonNull(completableTasks);
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/CompletableTaskList.java]: This one can use requireNonNull(), but no big deal.

File: src/main/java/seedu/address/model/project/ParticipantList.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public ParticipantList() {}
+
+    /**
+     * Constructs a {@code ParticipantList}.
+     *
+     * @param participants A list of {@code Person}.
+     */
+    public ParticipantList(List<Person> participants) {
+        requireAllNonNull(participants);
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/ParticipantList.java]: This one can use requireNonNull(), but no big deal.

File: src/main/java/seedu/address/model/project/UniqueProjectList.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.project.exceptions.DuplicateProjectException;
+import seedu.address.model.project.exceptions.ProjectNotFoundException;
+
+/**
+ * A list of projects that enforces uniqueness between its elements and does not allow nulls.
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/UniqueProjectList.java:15]: I love how your comments are so detailed btw! :)

File: src/main/java/seedu/address/model/project/UniqueProjectList.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+     */
+    public void setProject(Project target, Project editedProject) {
+        requireAllNonNull(target, editedProject);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new ProjectNotFoundException();
+        }
+
+        if (!target.isSameProject(editedProject) && contains(editedProject)) {
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/UniqueProjectList.java:66]: Should it be target.isSameProject(edited) && contains(editedProject) without the ! in front of target.isSameProject(edited)?

File: src/main/java/seedu/address/model/project/exceptions/ProjectNotFoundException.java [hunk 12]
```diff
@@ -0,0 +1,6 @@
+package seedu.address.model.project.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified project.
+ */
+public class ProjectNotFoundException extends RuntimeException {}
```
Comment (vevek) [inline src/main/java/seedu/address/model/project/exceptions/ProjectNotFoundException.java:6]: Nothing implemented here

**Review by vevek — DISMISSED** [review]

**Comment by vevek** [comment]
Might need to check the things brought up in the comments. Accidentally approved and left out this comment.

File: src/main/java/seedu/address/model/project/UniqueProjectList.java (hunk 10 above)
Comment (samuelfangjw) [reply src/main/java/seedu/address/model/project/UniqueProjectList.java:15]: Not my comments, copied from person haha

File: src/main/java/seedu/address/model/project/UniqueProjectList.java (hunk 11 above)
Comment (samuelfangjw) [reply src/main/java/seedu/address/model/project/UniqueProjectList.java:66]: can confirm no issue here, if target and edited are not the same, and current list already contains new edited project, then the project is duplicated.

File: src/main/java/seedu/address/model/project/exceptions/ProjectNotFoundException.java (hunk 12 above)
Comment (samuelfangjw) [reply src/main/java/seedu/address/model/project/exceptions/ProjectNotFoundException.java:6]: This was copied over from person class. Has same functionality as RuntimeException.

**Review by vevek — APPROVED** [review]
LGTM!

**Pull request merged**